### PR TITLE
[Grouping&Mapping] accessToken parameter now takes a callback fn

### PR DESCRIPTION
**Pull request by Guillar1**
accessToken now takes a callback function.

### CODE REVIEW

File: packages/itwin/grouping-mapping-widget/src/widget/GroupingMappingWidget.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/GroupingMappingWidget.tsx:28]: Since you are making a breaking change here, maybe change all these parameters to a single optional params object? It's pretty annoying when I want to provide only the `client` - need to pass 2 `undefined` values before I can pass in the client.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx:353]: Where are the errors handled?

File: packages/itwin/grouping-mapping-widget/src/widget/GroupingMappingWidget.tsx
Comment (MindaugasButkus) [reply packages/itwin/grouping-mapping-widget/src/widget/GroupingMappingWidget.tsx:28]: Maybe `GroupingMappingProps` could be reused here.

Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/GroupingMappingWidget.tsx:28]: Good idea

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx:353]: Oh, forgot to go back to this.

File: packages/itwin/grouping-mapping-widget/src/widget/GroupingMappingWidget.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/GroupingMappingWidget.tsx]: Converted to a single object params.

File: packages/itwin/grouping-mapping-widget/src/widget/GroupingMappingWidget.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/GroupingMappingWidget.tsx:28]: I don't think it's worth to deconstruct props here since later the same props object is rebuilt. If this was not reconstructed then it could be used like this: `<GroupingMapping {...this._props} />`.

Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/GroupingMappingWidget.tsx]: I agree. Changed it to pass the object directly.

Comment (MindaugasButkus) [reply packages/itwin/grouping-mapping-widget/src/widget/GroupingMappingWidget.tsx]: It could be simplified even further - the separate field declaration is not needed, there is a shorthand for constructors:
`constructor(private readonly _props: GroupingMappingProps = {}) {}`

Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/GroupingMappingWidget.tsx]: Oh right completely forgot about that shorthand. Thanks!

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupingMapping.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/GroupingMapping.tsx:48]: I think this might cause excessive re-renders when no `getAccessToken` is provided because each time a new function (lambda expression) will be created. Do you know if multiple instances of this lambda will have reference equality? If not, maybe a non-lambda function would have reference equality (non exported function could be defined in the same file).

File: packages/itwin/grouping-mapping-widget/src/widget/components/SelectProject.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/SelectProject.tsx:59]: Yup.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupingMapping.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupingMapping.tsx:48]: You are correct. The lambda will not have reference equality per render. I wrapped it in a useCallback. I also realized that the array dependencies could be more specific.

File: packages/itwin/grouping-mapping-widget/src/widget/components/SelectIModel.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/SelectIModel.tsx:33]: Just a note about `useMemo`: https://reactjs.org/docs/hooks-reference.html#usememo
`You may rely on useMemo as a performance optimization, not as a semantic guarantee.`

This place looks like you are relying on `useMemo` as a semantic guarantee to prevent re-renderings.

File: packages/itwin/grouping-mapping-widget/src/widget/components/SelectIModel.tsx
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/SelectIModel.tsx:52]: There is a chance that an expired `accessToken` will be passed into `IModelGrid`. New `accessToken` will only be fetched if `getAccessToken` function changes, not when the token expires. It would be best if the `IModelGrid` component would accept a function to get the token instead of the token itself. Chance of the "expired" `accessToken` issue occuring is relatively low, so we can leave this one for another PR.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupingMapping.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupingMapping.tsx]: I'll put the object in the value in a useState. Earlier, nothing was using the object as a dependency within the widget. Only the values within. Even if the object was recreated the properties would be unchanged and no rerender would happen. But I don't mind being safe about it.

File: packages/itwin/grouping-mapping-widget/src/widget/components/SelectIModel.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/SelectIModel.tsx]: Removed useMemo. I vaguely remember people screaming about this.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupingMapping.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupingMapping.tsx]: I put the object in the value in a useState. Nevermind my earlier comment.

**Comment by Guillar1** [comment]
@aruniverse I've added @MindaugasButkus to codeowners. Needs approval.

File: packages/itwin/grouping-mapping-widget/src/widget/components/SelectIModel.tsx
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/widget/components/SelectIModel.tsx]: Haha. It wasn't directed to me. There was some discussion on the documentation being a bit contradictory. 
https://reactjs.org/docs/hooks-faq.html#what-can-i-do-if-my-effect-dependencies-change-too-often
`As a last resort, you can add a function to effect dependencies but wrap its definition into the` [useCallback](https://reactjs.org/docs/hooks-reference.html#usecallback)` Hook. This ensures it doesn’t change on every render unless its own dependencies also change:`

In the `useCallback` docs it says `useCallback(fn, deps) is equivalent to useMemo(() => fn, deps).` So inherently its also subject to future cache busting. So it is a little confusing to suggest this pattern.

**Review by aruniverse — APPROVED** [review]

**Comment by aruniverse** [comment]
please have @MindaugasButkus approve before merging

**Review by MindaugasButkus — APPROVED** [review]
Approved.

**Pull request merged**
